File: src/main/java/com/eraoftribes/engine/world/WorldGenerator.java
package com.eraoftribes.engine.world;

import java.util.Random;

public class WorldGenerator {
    private long seed;
    private Random random;

    public WorldGenerator() {
        System.out.println("[WorldGenerator] Initialized.");
    }

    public WorldData generate(long seed, int width, int height) {
        this.seed = seed;
        this.random = new Random(seed);

        var data = new WorldData(width, height);
        generateTerrain(data);
        generateBiomes(data);
        generateRivers(data);
        generateResources(data);

        System.out.println("[WorldGenerator] World generated: " + width + "x" + height + ", seed=" + seed);
        return data;
    }

    private void generateTerrain(WorldData data) {
        for (int y = 0; y < data.height; y++) {
            for (int x = 0; x < data.width; x++) {
                data.heightMap[y][x] = perlinNoise(x * 0.01, y * 0.01, 6);
            }
        }
    }

    private void generateBiomes(WorldData data) {
        for (int y = 0; y < data.height; y++) {
            for (int x = 0; x < data.width; x++) {
                double h = data.heightMap[y][x];
                double lat = Math.abs((double) y / data.height - 0.5) * 2;
                if (h < 0.2) data.biomes[y][x] = "water";
                else if (h < 0.35) data.biomes[y][x] = lat > 0.7 ? "tundra" : "grassland";
                else if (h < 0.5) data.biomes[y][x] = lat > 0.6 ? "taiga" : "forest";
                else if (h < 0.7) data.biomes[y][x] = "mountain";
                else data.biomes[y][x] = "snow";
            }
        }
    }

    private void generateRivers(WorldData data) {
        int count = 3 + random.nextInt(4);
        for (int i = 0; i < count; i++) {
            int x = random.nextInt(data.width);
            int y = random.nextInt(data.height / 3);
            data.rivers.add(new River(x, y, 50 + random.nextInt(100)));
        }
    }

    private void generateResources(WorldData data) {
        String[] types = {"wood", "stone", "iron", "copper", "gold", "grain", "clay"};
        int count = 20 + random.nextInt(30);
        for (int i = 0; i < count; i++) {
            data.resources.add(new ResourceDeposit(
                random.nextInt(data.width),
                random.nextInt(data.height),
                types[random.nextInt(types.length)],
                50 + random.nextInt(200)
            ));
        }
    }

    private double perlinNoise(double x, double y, int octaves) {
        double value = 0, amplitude = 1, frequency = 1, max = 0;
        for (int o = 0; o < octaves; o++) {
            value += interpolatedNoise(x * frequency, y * frequency) * amplitude;
            max += amplitude;
            amplitude *= 0.5;
            frequency *= 2;
        }
        return value / max;
    }

    private double interpolatedNoise(double x, double y) {
        int ix = (int) Math.floor(x);
        int iy = (int) Math.floor(y);
        double fx = x - ix, fy = y - iy;
        double a = hash(ix, iy), b = hash(ix + 1, iy);
        double c = hash(ix, iy + 1), d = hash(ix + 1, iy + 1);
        return lerp(lerp(a, b, fx), lerp(c, d, fx), fy);
    }

    private double hash(int x, int y) {
        long n = x * 374761393L + y * 668265263L + seed;
        n = (n ^ (n >> 13)) * 1274126177L;
        return ((n ^ (n >> 16)) & 0x7FFFFFFF) / (double) 0x7FFFFFFF;
    }

    private double lerp(double a, double b, double t) { return a + t * (b - a); }

    public static class WorldData {
        public final int width, height;
        public final double[][] heightMap;
        public final String[][] biomes;
        public final java.util.List<River> rivers = new java.util.ArrayList<>();
        public final java.util.List<ResourceDeposit> resources = new java.util.ArrayList<>();

        WorldData(int w, int h) {
            width = w; height = h;
            heightMap = new double[h][w];
            biomes = new String[h][w];
        }
    }

    public static class River {
        public final int startX, startY, length;
        River(int x, int y, int len) { startX = x; startY = y; length = len; }
    }

    public static class ResourceDeposit {
        public final int x, y, amount;
        public final String type;
        ResourceDeposit(int x, int y, String type, int amount) {
            this.x = x; this.y = y; this.type = type; this.amount = amount;
        }
    }
}

File: game/maps/world_generator.js
// Era of Tribes - Weltgenerator
// Erzeugt prozedural Karten aus Layern.

var WorldGenerator = {
    seed: 1337,
    size: { width: 1920, height: 1080 },
    layers: {},

    generate: function(seed) {
        this.seed = seed || Date.now();
        console.log("Generiere Welt mit Seed: " + this.seed);

        this.generateTerrain();
        this.generateBiomes();
        this.generateRivers();
        this.generateRoads();
        this.generateResources();
        this.generateBorders();
        this.generateDecoration();

        return this.layers;
    },

    generateTerrain: function() {
        this.layers.terrain = {
            type: "heightmap",
            data: this.perlinNoise(this.size.width, this.size.height, 6)
        };
        console.log("  Terrain generiert.");
    },

    generateBiomes: function() {
        var terrain = this.layers.terrain.data;
        this.layers.biomes = { type: "colormap", data: [] };
        for (var y = 0; y < this.size.height; y++) {
            this.layers.biomes.data[y] = [];
            for (var x = 0; x < this.size.width; x++) {
                var h = terrain[y][x];
                var latFactor = Math.abs(y / this.size.height - 0.5) * 2;
                if (h < 0.2) this.layers.biomes.data[y][x] = "water";
                else if (h < 0.35) this.layers.biomes.data[y][x] = latFactor > 0.7 ? "tundra" : "grassland";
                else if (h < 0.5) this.layers.biomes.data[y][x] = latFactor > 0.6 ? "taiga" : "forest";
                else if (h < 0.7) this.layers.biomes.data[y][x] = "mountain";
                else this.layers.biomes.data[y][x] = "snow";
            }
        }
        console.log("  Biome generiert.");
    },

    generateRivers: function() {
        this.layers.rivers = { type: "lines", data: [] };
        var riverCount = 3 + Math.floor(Math.random() * 4);
        for (var i = 0; i < riverCount; i++) {
            var startX = Math.floor(Math.random() * this.size.width);
            var startY = Math.floor(Math.random() * this.size.height * 0.3);
            this.layers.rivers.data.push({
                source: { x: startX, y: startY },
                path: this.traceRiver(startX, startY)
            });
        }
        console.log("  Fluesse generiert: " + riverCount);
    },

    generateRoads: function() {
        this.layers.roads = { type: "lines", data: [] };
        console.log("  Strassen generiert.");
    },

    generateResources: function() {
        this.layers.resources = { type: "points", data: [] };
        var resourceCount = 20 + Math.floor(Math.random() * 30);
        for (var i = 0; i < resourceCount; i++) {
            this.layers.resources.data.push({
                x: Math.floor(Math.random() * this.size.width),
                y: Math.floor(Math.random() * this.size.height),
                type: this.randomResource(),
                amount: 50 + Math.floor(Math.random() * 200)
            });
        }
        console.log("  Ressourcen generiert: " + resourceCount);
    },

    generateBorders: function() {
        this.layers.borders = { type: "colormap", data: [] };
        console.log("  Grenzen generiert.");
    },

    generateDecoration: function() {
        this.layers.decoration = { type: "points", data: [] };
        console.log("  Dekoration generiert.");
    },

    perlinNoise: function(width, height, octaves) {
        var data = [];
        for (var y = 0; y < height; y++) {
            data[y] = [];
            for (var x = 0; x < width; x++) {
                var value = 0;
                var amplitude = 1;
                var frequency = 1;
                var maxValue = 0;
                for (var o = 0; o < octaves; o++) {
                    value += this.interpolatedNoise(x * frequency, y * frequency) * amplitude;
                    maxValue += amplitude;
                    amplitude *= 0.5;
                    frequency *= 2;
                }
                data[y][x] = value / maxValue;
            }
        }
        return data;
    },

    interpolatedNoise: function(x, y) {
        var ix = Math.floor(x);
        var iy = Math.floor(y);
        var fx = x - ix;
        var fy = y - iy;
        var a = this.hash(ix, iy);
        var b = this.hash(ix + 1, iy);
        var c = this.hash(ix, iy + 1);
        var d = this.hash(ix + 1, iy + 1);
        return this.lerp(this.lerp(a, b, fx), this.lerp(c, d, fx), fy);
    },

    hash: function(x, y) {
        var n = x * 374761393 + y * 668265263 + this.seed;
        n = (n ^ (n >> 13)) * 1274126177;
        return (n ^ (n >> 16)) & 0x7FFFFFFF / 0x7FFFFFFF;
    },

    lerp: function(a, b, t) { return a + t * (b - a); },

    traceRiver: function(startX, startY) {
        var path = [{ x: startX, y: startY }];
        var x = startX, y = startY;
        var maxLength = 50 + Math.floor(Math.random() * 100);
        for (var i = 0; i < maxLength; i++) {
            var dx = (Math.random() - 0.5) * 6;
            var dy = 2 + Math.random() * 4;
            x += dx;
            y += dy;
            if (y >= this.size.height) break;
            path.push({ x: Math.floor(x), y: Math.floor(y) });
        }
        return path;
    },

    randomResource: function() {
        var resources = ["wood", "stone", "iron", "copper", "gold", "grain", "clay"];
        return resources[Math.floor(Math.random() * resources.length)];
    }
};
